Src/WindowsServer/WindowsServer.Shared/Implementation/EnvironmentVariableMonitor.cs

```diff
@@ -0,0 +1,111 @@
+﻿namespace Microsoft.ApplicationInsights.WindowsServer.Implementation
+{
+    using System;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Threading;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+
+    /// <summary>
+    /// Utility to monitor the value of environment variables which may change 
+    /// during the run of an application. Checks the environment variables 
+    /// intermittently.
+    /// </summary>
+    internal abstract class EnvironmentVariableMonitor : IDisposable
+    {
+        // Environment variables tracked by this monitor.
+        protected readonly ConcurrentDictionary<string, string> CheckedValues;
+
+        // enabled flag primarily used during dispose
+        protected volatile bool isEnabled = true;
+
+        // how often we allow the code to re-check the environment
+        protected TimeSpan checkInterval;
+
+        // timer object that will periodically update the environment variables
+        private readonly Timer environmentCheckTimer;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="EnvironmentVariableMonitor" /> class.
+        /// </summary>
+        public EnvironmentVariableMonitor(IEnumerable<string> envVars, TimeSpan checkInterval)
+        {
+            this.CheckedValues = new ConcurrentDictionary<string, string>();
+            this.checkInterval = checkInterval;
+
+            foreach (string varName in envVars)
+            {
+                this.CheckedValues.TryAdd(varName, Environment.GetEnvironmentVariable(varName));
+            }
+
+            this.environmentCheckTimer = new Timer(this.CheckVariablesIntermittent, null, this.checkInterval, TimeSpan.FromMilliseconds(-1));
```

Comment:
If `this.checkInterval` can ever be 0 (i hope not), then there is a race condition possible - `this.CheckVariablesIntermittent` can be executed in its entirety before the assignment above happens, so `this.environmentCheckTimer()` will still be null. The last thing `this.CheckVariablesIntermittent()` does is to change `this.environmentCheckTimer` that may cause NullReference.
If I remember correctly, the right pattern was to create `Timer` like `Timer(callback, null, Timeout.Infinite, Timeout.Infinite)` and immediately follow up with `timer.Change(this.checkInterval, TimeoutInfinite)`.
Again, I think this only applies if you may have really-really low value as `this.checkInterval`.

Reply:
Ok, I wasn't aware of that possibility. I will make the change to a 2-step initialization here.

Reply:
I think I will enforce a minimum check interval as well - 5 seconds.